src/tools/rbd_mirror/PoolWatcher.cc

```diff
@@ -400,64 +443,65 @@ void PoolWatcher<I>::notify_listener() {
   dout(10) << dendl;
 
   std::string mirror_uuid;
-  ImageIds added_image_ids;
-  ImageIds removed_image_ids;
-  {
-    std::lock_guard locker{m_lock};
-    ceph_assert(m_notify_listener_in_progress);
-  }
-
-  if (!removed_image_ids.empty()) {
-    m_listener.handle_update(mirror_uuid, {}, std::move(removed_image_ids));
-    removed_image_ids.clear();
-  }
-
+  MirrorEntities added_entities;
+  MirrorEntities removed_entities;
   {
     std::lock_guard locker{m_lock};
     ceph_assert(m_notify_listener_in_progress);
 
     // if the watch failed while we didn't own the lock, we are going
     // to need to perform a full refresh
-    if (m_image_ids_invalid) {
+    if (m_entities_invalid) {
       m_notify_listener_in_progress = false;
       return;
     }
 
-    // merge add/remove notifications into pending set (a given image
+    // merge add/remove notifications into pending set (a given entity
     // can only be in one set or another)
-    for (auto &image_id : m_pending_removed_image_ids) {
-      dout(20) << "image_id=" << image_id << dendl;
-      m_pending_image_ids.erase(image_id);
+    for (auto &[entity, id] : m_pending_removed_entities) {
+      dout(20) << "removed pending entity=" << entity << dendl;
+      m_pending_entities.erase(entity);
+      auto it = m_entities.find(entity);
+      if (it != m_entities.end()) {
+        m_entities.erase(entity);
+        m_entities.insert({entity, id});
+      }
     }
+    m_pending_removed_entities.clear();
 
-    for (auto &image_id : m_pending_added_image_ids) {
-      dout(20) << "image_id=" << image_id << dendl;
-      m_pending_image_ids.erase(image_id);
-      m_pending_image_ids.insert(image_id);
+    for (auto &[entity, id] : m_pending_added_entities) {
+      dout(20) << "added pending entity=" << entity << dendl;
```

Comment:
```suggestion
      dout(20) << "added pending entity : " << entity << dendl;
```

The log message currently looks like this:
notify_listener: added pending entity=type=Group, global_id=de6a226b-102d-44c4-9405-011b8acc3aa1, count=1

The same comment holds for the log for removed entities.

Reply:
The other option is :
notify_listener: added pending entity={type=Group, global_id=de6a226b-102d-44c4-9405-011b8acc3aa1, count=1}